<a href="https://colab.research.google.com/github/MuhammadHelmyOmar/Hadith/blob/main/MAHDDAT.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

- MAHADDAT: MAwdu HADith DATaset

# Loading Corpus (MAHADDAT + NAH Plus + LK)

In [1]:
import pandas as pd

## Train Set

In [2]:
train_url = 'https://raw.githubusercontent.com/KamelGaanoun/MHDetection/main/Corpus/trainFinal.csv'

train = pd.read_csv(train_url)
print("CSV file loaded successfully!")
# Display the first few rows of the DataFrame
train.head()

CSV file loaded successfully!


Unnamed: 0                                              Matan  Degree
0       19036                    لقد رأيتني وأنا ثلث الإسلام،‏.‏       0
1       17209   هريرة، أن أعرابيا، أتى النبي صلى الله عليه وس...       0
2       11245   عائشة، قالت كنت أفتل قلائد هدى رسول الله صلى ...       0
3        2827  قال النبي صلى الله عليه وسلم ‏ ‏ إني فرطكم على...       0
4       23616  عن النبي صلى الله عليه وسلم قال أقيموا الركوع ...       0

In [3]:
print(len(train))
print(train.Degree.unique())

21248
[0 1]


In [26]:
train.Degree.value_counts()

Degree
0    19286
1     1962
Name: count, dtype: int64

## Test Set

In [4]:
test_url = 'https://raw.githubusercontent.com/KamelGaanoun/MHDetection/main/Corpus/testFinal.csv'

test = pd.read_csv(test_url)
print("CSV file loaded successfully!")
# Display the first few rows of the DataFrame
test.head()

CSV file loaded successfully!


Unnamed: 0                                              Matan  Degree
0       12957   أن رسول الله صلى الله عليه وسلم احتجم وهو محر...       0
1         593  أن رسول الله صلى الله عليه وسلم كان يسأل في مر...       0
2       13256  قال كنا نحاقل الأرض على عهد رسول الله صلى الله...       0
3         519   أن النبي صلى الله عليه وسلم كان إذا وضع الميت...       0
4       21177                    تختموا بالياقوت فإنه ينفي الفقر       1

In [5]:
print(len(test))
print(test.Degree.unique())

2657
[0 1]


In [24]:
test.Degree.value_counts()

Degree
0    2412
1     245
Name: count, dtype: int64

## Dev Set

In [6]:
dev_url = 'https://raw.githubusercontent.com/KamelGaanoun/MHDetection/main/Corpus/devFinal.csv'

dev = pd.read_csv(dev_url)
print("CSV file loaded successfully!")
# Display the first few rows of the DataFrame
dev.head()

CSV file loaded successfully!


Unnamed: 0                                              Matan  Degree
0       11697  قال كان رسول الله صلى الله عليه وسلم بالبقيع ف...       0
1       23046   هريرة، أن رسول الله صلى الله عليه وسلم نعى لل...       0
2       21715  أنه سمع عمران بن حصين، يحدث عن النبي صلى الله ...       0
3       16238  قال كان رسول الله صلى الله عليه وسلم يصلي بنا ...       0
4       15667  كان الناس يتحرون بهداياهم يوم عائشة قالت عائشة...       0

In [7]:
print(len(dev))
print(dev.Degree.unique())

2656
[0 1]


In [25]:
dev.Degree.value_counts()

Degree
0    2411
1     245
Name: count, dtype: int64

# Loading MAHADDAT Only

In [8]:
mahaddat_url = 'https://raw.githubusercontent.com/KamelGaanoun/MHDetection/main/MAHADDAT/MAHADDAT.csv'

mahaddat = pd.read_csv(mahaddat_url)
print("CSV file loaded successfully!")
# Display the first few rows of the DataFrame
mahaddat.head()

CSV file loaded successfully!


Book No.                                        Full Hadith  \
0   N1  16   رقم الحديث: 15   (حديث مرفوع)    أخبرنا  محمد...   
1   N1  18   رقم الحديث: 17   (حديث مرفوع)    أخبرنا  طاهر...   
2   N1  20   رقم الحديث: 19   (حديث مرفوع)    أخبرنا  أبو ...   
3   N1  30   رقم الحديث: 29   (حديث مرفوع)    أخبرنا  أبو ...   
4   N1  35   رقم الحديث: 34   (حديث مرفوع)    أخبرنا  أبو ...   

                                               Isnad  \
0   أخبرنا  محمد بن الحسين بن محمد الواعظ  ، أخبر...   
1  أخبرنا  طاهر بن الفضل بن أحمد الحافظ  ، قال : ...   
2   أخبرنا  أبو العلاء الأديب   ، أخبرنا  علي بن ...   
3   أخبرنا  أبو القاسم حمد بن عبد الواحد بن إسماع...   
4  أخبرنا  أبو العلاء نصر بن أحمد الأديب   ، أخبر...   

                                               Matan  \
0                         الإيمان لا يزيد ولا ينقص "   
1  أن وفد ثقيف جاءوا إلى النبي صلى الله عليه وسلم...   
2   من زعم أن الإيمان يزيد وينقص ، فزيادته نفاق ،...   
3   من لم يميز ثلاثة فليس له في الجماعة نصيب : من...   
4  صنفان من أمتي لا تنالهما شفاعتي : المرجئة والق...   

                                   Author's Comments Degree   Authenticity  \
0  وهو من موضوعات أحمد بن عبد الله الجويباري ، وأ...  Mawdu  non-authentic   
1  وهو من موضوعات أبي مطيع البلخي ، وأبو مطيع ذا ...  Mawdu  non-authentic   
2  وهو من موضوعات محمد بن القاسم الطايكاني ، ومحم...  Mawdu  non-authentic   
3   هذا حديث باطل لا شك فيه ، وأحمد بن عبد الله ا...  Mawdu  non-authentic   
4  وفي إسناده ظلمات ، منها سلمة بن وردان ، قال : ...  Mawdu  non-authentic   

     Topic  
0  الإيمان  
1  الإيمان  
2  الإيمان  
3  الإيمان  
4  الإيمان

In [9]:
print(len(mahaddat))
print(mahaddat.columns)
print(mahaddat.Book.unique())

2452
Index(['Book', 'No.', 'Full Hadith', 'Isnad', 'Matan', 'Author's Comments',
       'Degree', 'Authenticity', 'Topic'],
      dtype='object')
['N1' 'N2' 'N3_1' 'N3_2' 'N5' 'N6' nan]


- What is nan book?!

In [10]:
nan_book_rows = mahaddat[mahaddat['Book'].isna()]
print("Rows with 'Book' as NaN:")
print(len(nan_book_rows))
nan_book_rows.head()

Rows with 'Book' as NaN:
395


Book  No. Full Hadith Isnad  \
2057  NaN  NaN         NaN   NaN   
2058  NaN  NaN         NaN   NaN   
2059  NaN  NaN         NaN   NaN   
2060  NaN  NaN         NaN   NaN   
2061  NaN  NaN         NaN   NaN   

                                                  Matan Author's Comments  \
2057   قَالَ مَنْ قَالَ إِذَا أَصْبَحَ وَإِذَا أَمْس...               NaN   
2058   أَنَّ النَّبِيَّ صلى الله عليه وسلم نَهَى أَن...               NaN   
2059  أَنَّ عَمْرَةَ بِنْتَ الْجَوْنِ، تَعَوَّذَتْ م...               NaN   
2060  أَمَرَ رَسُولُ اللَّهِ صلى الله عليه وسلم الأَ...               NaN   
2061  قَالَ كُنَّا عِنْدَ رَسُولِ اللَّهِ صلى الله ع...               NaN   

     Degree Authenticity Topic  
2057  Mawdu          NaN   NaN  
2058  Mawdu          NaN   NaN  
2059  Mawdu          NaN   NaN  
2060  Mawdu          NaN   NaN  
2061  Mawdu          NaN   NaN

- nan book are the mawdu hadiths from LK (29) plus the scrapped mawdu hadiths from dorar (366).

In [11]:
print(mahaddat["No."].unique())
print(len(mahaddat["No."].unique()))
print(mahaddat.Degree.unique())
print(mahaddat.Authenticity.unique())

['16' '18' '20' ... '938_2' '1375_4' nan]
1669
['Mawdu']
['non-authentic' 'non_authentic' 'unknown' 'authentic' nan]


- It is strange that all hadiths are mawdu, but still some hadiths are considered authentic!

In [35]:
mahaddat.Authenticity.value_counts()

Authenticity
non-authentic    1567
non_authentic     483
unknown             4
authentic           1
Name: count, dtype: int64

In [12]:
nan_auth_rows = mahaddat[mahaddat['Authenticity'].isna()]
print("Rows with 'Authenticity' as NaN:")
print(len(nan_auth_rows))
nan_auth_rows.head()

Rows with 'Authenticity' as NaN:
397


Book    No.                                        Full Hadith  \
1274  N3_1  276_6                                                NaN   
1608  N3_2    887  (الخطيب) أنبأنا القاضي أبو العلاء محمد بن علي ...   
2057   NaN    NaN                                                NaN   
2058   NaN    NaN                                                NaN   
2059   NaN    NaN                                                NaN   

                                                  Isnad  \
1274                                                NaN   
1608  الخطيب) أنبأنا القاضي أبو العلاء محمد بن علي ب...   
2057                                                NaN   
2058                                                NaN   
2059                                                NaN   

                                                  Matan  \
1274    من قرأ ليلة الجمعة حم الدخان ويس أصبح مغفورا له   
1608   يأتي على جهنم يوماً ما فيها من بني آدم أحد تح...   
2057   قَالَ مَنْ قَالَ إِذَا أَصْبَحَ وَإِذَا أَمْس...   
2058   أَنَّ النَّبِيَّ صلى الله عليه وسلم نَهَى أَن...   
2059  أَنَّ عَمْرَةَ بِنْتَ الْجَوْنِ، تَعَوَّذَتْ م...   

                     Author's Comments Degree Authenticity  \
1274                               NaN  Mawdu          NaN   
1608   موضوع: جعفر هو ابن الزبير متروك  Mawdu          NaN   
2057                               NaN  Mawdu          NaN   
2058                               NaN  Mawdu          NaN   
2059                               NaN  Mawdu          NaN   

                        Topic  
1274  ‏كتاب الأنبياء والقدماء  
1608             فوائد متفرقة  
2057                      NaN  
2058                      NaN  
2059                      NaN

In [15]:
print(f"Number of topics = {len(mahaddat.Topic.unique())}")
# print(mahaddat.Topic.unique())

Number of topics = 78


# Loading NAH Plus Only

In [16]:
NAHPlus_url = 'https://raw.githubusercontent.com/KamelGaanoun/MHDetection/main/NAH_plus/NAH_plus.csv'

NAHPlus = pd.read_csv(NAHPlus_url)
print("CSV file loaded successfully!")
# Display the first few rows of the DataFrame
NAHPlus.head()

CSV file loaded successfully!


Unnamed: 0.1 Book No.                                        Full Hadith  \
0             0   N1  16   رقم الحديث: 15   (حديث مرفوع)    أخبرنا  محمد...   
1             1   N1  18   رقم الحديث: 17   (حديث مرفوع)    أخبرنا  طاهر...   
2             2   N1  20   رقم الحديث: 19   (حديث مرفوع)    أخبرنا  أبو ...   
3             3   N1  30   رقم الحديث: 29   (حديث مرفوع)    أخبرنا  أبو ...   
4             4   N1  35   رقم الحديث: 34   (حديث مرفوع)    أخبرنا  أبو ...   

                                               Isnad  \
0   أخبرنا  محمد بن الحسين بن محمد الواعظ  ، أخبر...   
1  أخبرنا  طاهر بن الفضل بن أحمد الحافظ  ، قال : ...   
2   أخبرنا  أبو العلاء الأديب   ، أخبرنا  علي بن ...   
3   أخبرنا  أبو القاسم حمد بن عبد الواحد بن إسماع...   
4  أخبرنا  أبو العلاء نصر بن أحمد الأديب   ، أخبر...   

                                               Matan  \
0                         الإيمان لا يزيد ولا ينقص "   
1  أن وفد ثقيف جاءوا إلى النبي صلى الله عليه وسلم...   
2   من زعم أن الإيمان يزيد وينقص ، فزيادته نفاق ،...   
3   من لم يميز ثلاثة فليس له في الجماعة نصيب : من...   
4  صنفان من أمتي لا تنالهما شفاعتي : المرجئة والق...   

                                   Author's Comments Degree   Authenticity  \
0  وهو من موضوعات أحمد بن عبد الله الجويباري ، وأ...  Mawdu  non-authentic   
1  وهو من موضوعات أبي مطيع البلخي ، وأبو مطيع ذا ...  Mawdu  non-authentic   
2  وهو من موضوعات محمد بن القاسم الطايكاني ، ومحم...  Mawdu  non-authentic   
3   هذا حديث باطل لا شك فيه ، وأحمد بن عبد الله ا...  Mawdu  non-authentic   
4  وفي إسناده ظلمات ، منها سلمة بن وردان ، قال : ...  Mawdu  non-authentic   

     Topic  Unnamed: 0  
0  الإيمان         NaN  
1  الإيمان         NaN  
2  الإيمان         NaN  
3  الإيمان         NaN  
4  الإيمان         NaN

In [17]:
print(len(NAHPlus))
print(NAHPlus.Book.unique())
print(len(NAHPlus["No."].unique()))

3660
['N1' 'N2' 'N3_1' 'N3_2' 'N5' 'N6' 'N4']
2496


In [18]:
# print(NAHPlus.Degree.unique())
print(f'Number of Degree labels = {len(NAHPlus.Degree.unique())}')
print(NAHPlus.Authenticity.unique())

Number of Degree labels = 94
['non-authentic' 'non_authentic' 'unknown' nan 'authentic/non-authentic'
 'authentic']


In [34]:
NAHPlus.Degree.value_counts()

Degree
Mawdu                1191
Daif                 1015
موضوع                 509
Unknown               480
باطل                  136
                     ... 
لعن الله من وضعه        1
[منكر]                  1
شبه موضوع               1
قول مجاهد               1
بهذا الإسناد منكر       1
Name: count, Length: 94, dtype: int64

In [19]:
nan_auth_rows = NAHPlus[NAHPlus['Authenticity'].isna()]
print("Rows with 'Authenticity' as NaN:")
print(len(nan_auth_rows))
nan_auth_rows.head()

Rows with 'Authenticity' as NaN:
6


Unnamed: 0.1  Book    No.  \
1344          1344    N1    436   
1581          1581  N3_1    349   
1602          1602  N3_1  476_2   
1617          1617  N3_1    567   
1619          1619  N3_1    574   

                                            Full Hadith  \
1344   رقم الحديث: 426   (حديث مرفوع)    أخبرنا  الح...   
1581  (الخطيب) أنبأنا أبو الفرج عبدالوهاب بن الحسين ...   
1602                                                NaN   
1617  (العقيلي) حدثنا داود بن محمد حدثنا أبو إبراهيم...   
1619  (‏ حدثت) عن عبدالرحمن بن عوف بن محمد حدثنا أحم...   

                                                  Isnad  \
1344  أخبرنا  الحسين بن علي  ، أخبرنا  علي بن محمد  ...   
1581  أنبأنا أبو الفرج عبدالوهاب بن الحسين بن عمر بن...   
1602   (أبو بكر) الشافعي حدثنتي سماتة بنت حمدان بن م...   
1617  العقيلي) حدثنا داود بن محمد حدثنا أبو إبراهيم ...   
1619   حدثت) عن عبدالرحمن بن عوف بن محمد حدثنا أحمد ...   

                                                  Matan  \
1344   لما مات إبراهيم ابن النبي صلى الله عليه وسلم ...   
1581  لما نزلت إذا جاء نصر اللّه والفتح جاء العباس إ...   
1602   لما مات ولدي من خديجة أوحى اللّه إلي أن أمسك ...   
1617   رأى أبو هريرة رجلاً فأعجبه هيئته فقال ممن أنت...   
1619  يكون في أمتي رجل يقال له محمد بن إدريس أضر على...   

                                      Author's Comments Degree Authenticity  \
1344                                                NaN   Daif          NaN   
1581  عمر كذاب (قلت) قال في الميزان هذا الحديث ليس ب...   Daif          NaN   
1602   موضوع: عمر وابن زياد كذاب يضع (قلت) قال في ال...   Daif          NaN   
1617   لا أصل له عبدالرحمن متروك وقال أبو داود كذاب ...   Daif          NaN   
1619   موضوع: وضعه المأمون أو الجويباري وذكر الحاكم ...   Daif          NaN   

             Topic  Unnamed: 0  
1344        الصلاة         NaN  
1581  كتاب المناقب         NaN  
1602  كتاب المناقب         NaN  
1617  كتاب المناقب         NaN  
1619  كتاب المناقب         NaN

In [36]:
NAHPlus.Authenticity.value_counts()

Authenticity
non-authentic              2846
non_authentic               648
unknown                     156
authentic                     3
authentic/non-authentic       1
Name: count, dtype: int64

- I don't understand yet what is meant by authenticity column here!

In [23]:
print(f"Number of topics = {len(NAHPlus.Topic.unique())}")
# print(mahaddat.Topic.unique())

Number of topics = 122


# Next Task

- Check the large corpus filtering
- Explore the LK for other authenticity level
- Use pandas distribution function to check noise in qunique values
- How many hadiths you can't map to existed authenticity level (authentic-good-weak-fabricated)
(صحيح-حسن-ضعيف-موضوع)
- Check how MAHADDAT filtered the mawdu from LK

# QA

## Large Corpus Creation Process (page 4)
- It is unbalanced for Mawdu and authentic Hadiths (authentic is the majority).
- Cleaning Phase:
  - Cleaned Matn and degree of authenticity columns.
  - For LK (34,088):
    - Using Dorar to check Hadiths when Arabic and English grades differ (2).
    - Removing Hadiths with empty Arabic Matn (826).
    - Removing Hadiths with empty English and Arabic grades (380).
    - Removing Hadiths with no English grade and an Arabic grade that we were not able to classify as authentic or not (32).
    - Keep the following grades:
      - Sahih - Authentic
      - Daif - Weak
      - Hassan - Good
      - Hassan - Sahih
      - Mawdu - Fabricated
      - Munkar
      - Remove the rest (708)
    - Remaining = 32,142
  - For NAH (7,363):
    - Scraping the degree from Hdith website for Hadiths without degree (3,552).
      - 900 Hadiths are detected as Mawdu
    - Removal of Hadiths without Matn (1,246).
    - Removal of authentic Hadiths (359).
    - Grouping Hadiths with degrees meaning Mawdu and leave the remaining as Daif, so grades:
      - Mawdu
      - Daif
    - Remaining = 5758
- NAHPlus contains 3660 Hadiths.
  - The enhanced version of NAH after cleaning and scraping.
  - Ony non-authentic
- Filter Phase:
  - Extract Mawdu from NAHPlus and LK(29).
  - Add 366 MHs scraped from Dorar.
  - Removing Duplicates.
  - MAHADDAT is the result and contains 2,452 MHs.
  - The final large corpus is a mix of:
    - MAHADDAT
    - SHs from LK
  - Duplicates are removed.
  - Hadiths with a Matn indicating their authenticity degree were identified and removed from the LK corpus.
  - Hadiths containing the words authentic (صحيح) or good (حسن) are removed.
  - Elemnate the starting expressions (قال رسول الله..إلخ) from LK.
- Final corpus contains 26,561 Hadiths.
  - 24,109 are Sahih
  - 2,452 are Mawdu
  - If we omit the extremes, the MHs have much lower words than the SHs.
  - For longer Hadiths, Mawdu ones can be three times longer than SHs, attaining a maximum of 5,608 words.
  - Built a list of stopwords specific to Hadiths, words like:
    - عز وجل: Almighty
    - الله: God
    - فقال: and he said
    - سمعت: I heard
    - النبي: The prophet
  


In [33]:
len(train)+len(test)+len(dev)

26561